cpp/benchmarks/groupby/group_max_multithreaded.cpp

```diff
@@ -0,0 +1,102 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include <benchmarks/common/generate_input.hpp>
+#include <benchmarks/fixture/benchmark_fixture.hpp>
+
+#include <cudf/detail/utilities/stream_pool.hpp>
+#include <cudf/groupby.hpp>
+#include <cudf/utilities/default_stream.hpp>
+#include <cudf/utilities/thread_pool.hpp>
+
+#include <nvbench/nvbench.cuh>
+
+template <typename Type>
+void bench_groupby_max_multithreaded(nvbench::state& state, nvbench::type_list<Type>)
+{
+  auto const cardinality      = static_cast<cudf::size_type>(state.get_int64("cardinality"));
+  auto const num_rows         = static_cast<cudf::size_type>(state.get_int64("num_rows"));
+  auto const null_probability = state.get_float64("null_probability");
+  auto const num_threads      = state.get_int64("num_threads");
+  auto const num_aggregations = state.get_int64("num_aggregations");
+
+  auto const keys = [&] {
+    data_profile const profile =
+      data_profile_builder()
+        .cardinality(cardinality)
+        .no_validity()
+        .distribution(cudf::type_to_id<int32_t>(), distribution_id::UNIFORM, 0, num_rows);
+    return create_random_column(cudf::type_to_id<int32_t>(), row_count{num_rows}, profile);
+  }();
+
+  auto const vals = [&] {
+    auto builder = data_profile_builder().cardinality(0).distribution(
+      cudf::type_to_id<Type>(), distribution_id::UNIFORM, 0, num_rows);
+    if (null_probability > 0) {
+      builder.null_probability(null_probability);
+    } else {
+      builder.no_validity();
+    }
+    return create_random_column(
+      cudf::type_to_id<Type>(), row_count{num_rows}, data_profile{builder});
+  }();
+
+  auto keys_view = keys->view();
+  auto gb_obj    = cudf::groupby::groupby(cudf::table_view({keys_view, keys_view, keys_view}));
+
+  auto streams = cudf::detail::fork_streams(cudf::get_default_stream(), num_threads);
+  cudf::detail::thread_pool threads(num_threads);
+
+  std::vector<std::vector<cudf::groupby::aggregation_request>> requests(num_threads);
+  for (auto& thread_requests : requests) {
+    for (int64_t j = 0; j < num_aggregations; j++) {
+      thread_requests.emplace_back();
+      thread_requests.back().values = vals->view();
+      thread_requests.back().aggregations.push_back(
+        cudf::make_max_aggregation<cudf::groupby_aggregation>());
+    }
+  }
+
+  auto const mem_stats_logger = cudf::memory_stats_logger();
+  state.exec(
+    nvbench::exec_tag::sync | nvbench::exec_tag::timer, [&](nvbench::launch& launch, auto& timer) {
+      auto perform_agg = [&](int64_t index) { gb_obj.aggregate(requests[index], streams[index]); };
+      for (int64_t i = 0; i < num_threads; ++i) {
+        threads.submit(perform_agg, i);
+      }
+      timer.start();
```

Comment:
Apologies if I missed this in a thread, but how come we don't pause/unpause threads, like we do in ORC and Parquet benchmarks?

Reply:
Please look at this thread https://github.com/rapidsai/cudf/pull/16154#discussion_r1665321274 mentioned above, which might be a better pattern for multiple streams.

Reply:
Couple of questions, based on the thread: 

1. Should timer start before submits?
2. Should we apply this pattern to ORC and Parquet multi-threaded benchmarks?

Reply:
1. IMO, timer should also take tasks submitting time into consideration (they might launch asynchronously and fire-off immediately).
2. I am not sure about that. @wence- or @PointKernel may have better answers :).

Reply:
> IMO, timer should also take tasks submitting time into consideration (they might launch asynchronously and fire-off immediately).

Let's do that, then. I think that's my last outstanding comment on the PR.

Reply:
> Should we apply this pattern to ORC and Parquet multi-threaded benchmarks?

I think so. We didn't encounter this issue with parquet/orc probably because those are more timely expensive than groupby thus the threading overhead is negligible.